[sw, dif_rstmgr] Add control and read APIs for alert info crash dump

silvestrst opened the pull request:
This change is a part of DIF Reset Manager follow-up changes described in #3454.

I will add the corresponding unit tests in the follow-up when the API for this change has been approved.

silvestrst commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+
+  uint32_t enabled = (state == kDifRstmgrToggleEnabled) ? 0x1 : 0x0;
+
+  // This will clobber the `ALERT_INFO_CTRL.INDEX` field. However, the index
+  // field is only relevant during the crash dump read operation, and is
+  // set by the caller and not the hardware, so it is safe to clobber it.
+  mmio_region_write32(base_addr, RSTMGR_ALERT_INFO_CTRL_REG_OFFSET, enabled);
```
I wonder if I should just write only to the bit `0` by using the `bitfield` library? It will involve couple additional calls, and Read-Write-Modify arrangement, but potentially makes the implementation more robust in case this register layout changes.

lenary replied:
I don't think this is an issue, but it would be good to get @tjaychen to confirm, and we may need a note about this in the documentation on this function in the header.

tjaychen replied:
yep this looks okay to me!

mcy commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+
+  mmio_region_t base_addr = handle->params.base_addr;
+
+  uint32_t enabled = (state == kDifRstmgrToggleEnabled) ? 0x1 : 0x0;
```
I wouldn't bother with the ternary. `? 1 : 0` is the same (from a readability perspective) as `? true : false`.

mcy commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+    return kDifRstmgrBadArg;
+  }
+
+  mmio_region_t base_addr = handle->params.base_addr;
```
Could you just inline this? If you really want to keep it, move it to the site of first use.

mcy commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+  uint32_t control_reg =
+      mmio_region_read32(base_addr, RSTMGR_ALERT_INFO_CTRL_REG_OFFSET);
+
+  control_reg = bitfield_field32_write(
+      control_reg,
+      (bitfield_field32_t){.mask = RSTMGR_ALERT_INFO_CTRL_INDEX_MASK,
+                           .index = RSTMGR_ALERT_INFO_CTRL_INDEX_OFFSET},
+      index);
```
If you include a trailing comma in the designated initializer, clang-format will wrap it in a different, more readable manner.

alphan replied:
IMHO either way is fine.

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_rstmgr_result_t dif_rstmgr_alert_info_set_enabled(
```
Should there be a `get` variant?

alphan replied:
Would it make sense to mention that this is reset when the system is reset (as opposed to the crash dump)? From hw spec:
> This field is automatically set to 0 upon system reset (even if rstmgr is not reset).

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+    const dif_rstmgr_t *handle, size_t *count);
+
+/**
+ * Alert info crash dump 32-bit value.
```
Is this value opaque? Is its 32-bit-ness part of its API? I think this should be reflected in the doc comment.

alphan replied:
I wondered this as well.

silvestrst replied:
Done.

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+ * the information after all type of resets except Power-On-Reset (POR).
+ *
+ * @param handle A Reset Manager handle.
+ * @param count The size of alert info crash dump.
```
This @param looks stale?

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+/**
+ * Reads a single alert info crash dump `value` at `index`.
+ *
+ * Alert info crash dump consists of one or more
+ * `dif_rstmgr_alert_info_dump_value_t`. The total count can be found through
+ * using the `dif_rstmgr_alert_info_dump_get_size`.
```
Could you proofread this? I think you dropped a `function` at the end of this paragraph, and the wording seems odd.

mcy commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+ * `dif_rstmgr_alert_info_dump_value_t`. The total count can be found through
+ * using the `dif_rstmgr_alert_info_dump_get_size`.
+ *
+ * The corresponding registers are always-on and are expected to retain
+ * the information after all type of resets except Power-On-Reset (POR).
```
Consider this phrasing that's a bit more SW-focused: "The crash dump is always retained after any kind of reset, except on POR."

silvestrst replied:
Done, thanks this was very helpful.

mcy commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+
+  mmio_region_t base_addr = handle->params.base_addr;
+
+  if (index >= rstmgr_alert_info_dump_get_size(base_addr)) {
+    return kDifRstmgrBadArg;
+  }
```
I'm not sure this is a good API. Note that `BadArg` is incorrect here, since this performs an MMIO operation.

I would go with expecting the caller to not pass an out-of-bounds length. Alternatively, you could move to a transactional iterator API where a caller begins a crash dump read and repeatedly calls `end()` to get the next word in the dump.

alphan replied:
> Note that BadArg is incorrect here

Well, there is an MMIO read but AFAICT it has no side-effects because the register is read-only.

lenary replied:
Sorry, we've tried to be clear here, the read itself is a side-effect.

For simplicitly, this should be `kDifRstmgrError` if you don't want to add another variant/enum.

alphan commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +159,67 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+  return kDifRstmgrOk;
+}
+
+dif_rstmgr_result_t dif_rstmgr_alert_info_dump_read(
+    const dif_rstmgr_t *handle, size_t index,
+    dif_rstmgr_alert_info_dump_value_t *value) {
```
Just a suggestion: If 
* `INDEX` in `ALERT_INFO_CTRL` is reset to 0 on system reset, and
* We don't need to support reading the same dump multiple times,

then, we can use `INDEX` in `ALERT_INFO_CTRL` to keep track of the read operation and have something like:
```
result_t dif_rstmgr_alert_info_dump_read(... *handle, ... *buffer, buffer_len, size_t *words_written)
```

But, this doesn't make a whole lot of sense if the only use-case of this API is to dump these values one by one over a serial interface. Something like this might be better:
```
// index >= cnt_avail?
result_t dif_rstmgr_alert_info_dump_is_empty(... *handle, ... *is_empty);
// read next dump word
result_t dif_rstmgr_alert_info_dump_read_next(... *handle, ... *dump_word)
```

silvestrst replied:
Actually, probably even better solution would be @lenary suggested `typedef struct dif_rstmgr_info_dump{type data[index_mask_size]}`. 

`result_t dif_rstmgr_alert_info_dump_read(... *handle, ... dif_rstmgr_info_dump_t *dump, size_t *words_written)`

alphan replied:
I agree, max size of the dump is pretty small. What do you think about putting the size, i.e. the actual number of words, in the struct as well?

lenary commented on sw/device/lib/dif/dif_rstmgr.h:
```diff
@@ -196,6 +210,58 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_get(
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_rstmgr_result_t dif_rstmgr_alert_info_dump_read(
+    const dif_rstmgr_t *handle, size_t index,
+    dif_rstmgr_alert_info_dump_value_t *value);
```
It seems there's a lot of suggestions for this API, and the like. 

I'd like to avoid a stateful cursor-based API, so I think the solution might just be to go with a single function that reads all the dump values into a buffer, which also says how many entries are in there when it's complete. I think we can bound the size of the buffer required because we know how many bits there are in the INDEX field used for accessing them. This feels nicer than a stateful API or re-reading the size field on every access to make sure the index we chose was in range.

silvestrst replied:
Going with @alphan [solution](https://github.com/lowRISC/opentitan/pull/3867/commits/b7d4b68f4a97a0c445df3881b948c1cc2c59e7a9#r508598310).

silvestrst reviewed (commented):
I am planning on fixing these review comments at some point this week.

silvestrst reviewed (commented):
Addressed all of the outstanding comments.

alphan commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -38,6 +38,11 @@ _Static_assert(
 
+// Make sure that the public alert info crash dump size matches the HW.
+_Static_assert(
+    DIF_RSTMGR_ALERT_INFO_MAX_SIZE == RSTMGR_ALERT_INFO_CTRL_INDEX_MASK,
```
I'm a little bit confused because both `INDEX` and `CNT_AVAIL` are 4 bits wide. Normally I would say size should be mask + 1, but it looks like max size is indeed 15 in this case and therefore max index is 14, which is a little bit odd. Apologies if this was discussed before but just wanted to make sure.

alphan commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +164,86 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+  // Read the entire alert info crash dump, one 32bit data segment at the time.
+  dump->size = rstmgr_alert_info_dump_get_size(base_addr);
+  for (int i = 0; i < dump->size; ++i) {
+    uint32_t control_reg =
+        mmio_region_read32(base_addr, RSTMGR_ALERT_INFO_CTRL_REG_OFFSET);
```
This can be moved out of the loop, right?

alphan commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +164,86 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+  return kDifRstmgrOk;
+}
+
+dif_rstmgr_result_t dif_rstmgr_alert_info_dump_get_size(
+    const dif_rstmgr_t *handle, size_t *count) {
+  if (handle == NULL || count == NULL) {
+    return kDifRstmgrBadArg;
+  }
+
+  mmio_region_t base_addr = handle->params.base_addr;
+
+  *count = rstmgr_alert_info_dump_get_size(base_addr);
+
+  return kDifRstmgrOk;
+}
```
Yes, I don't think this is needed anymore.

I was thinking if a `..._has_dump` would be useful but I don't think it is necessary because clients can infer this by calling `dif_rstmgr_alert_info_dump_read` and checking its size. Up to you.

alphan commented on sw/device/lib/dif/dif_rstmgr.c:
```diff
@@ -155,6 +164,86 @@ dif_rstmgr_result_t dif_rstmgr_reset_info_clear(const dif_rstmgr_t *handle) {
+  };
+
+  // Read the entire alert info crash dump, one 32bit data segment at the time.
+  dump->size = rstmgr_alert_info_dump_get_size(base_addr);
```
nit: You may want to inline this if it's not used anywhere else.

silvestrst commented:
**I am not sure that the following is a good idea, what do you guys think?**

`dif_rstmgr.h`:
```
...
/**                                                                                                                                                                                                                 
 * The maximal size of the alert crash info dump.                                                                                                                                                                   
 *                                                                                                                                                                                                                  
 * Note: must match the autogenerated register definition.                                                                                                                                                          
 */
#define DIF_RSTMGR_ALERT_INFO_MAX_SIZE 0xf
...
/**                                                                                                                                                                                                                 
 * Reset Manager possible reset information enumeration.                                                                                                                                                            
 *                                                                                                                                                                                                                  
 * Invariants are used to extract information encoded in                                                                                                                                                            
 * `dif_rstmgr_reset_info_bitfield_t`, which means that the values must                                                                                                                                             
 * correspond                                                                                                                                                                                                       
 * to the individual bits (0x1, 0x2, 0x4, ..., 0x80000000).                                                                                                                                                         
 *                                                                                                                                                                                                                  
 * Note: these must match the autogenerated register definitions.                                                                                                                                                   
 */
typedef enum dif_rstmgr_reset_info {
  /**                                                                                                                                                                                                               
   * Device has reset due to power up.                                                                                                                                                                              
   */
  kDifRstmgrResetInfoPor = 0x1,
  /**                                                                                                                                                                                                               
   * Device has reset due to lowe power exit.                                                                                                                                                                       
   */
  kDifRstmgrResetInfoLowPowerExit = (0x1 << 1),
  /**                                                                                                                                                                                                               
   * Device has reset due to non-debug-module request.                                                                                                                                                              
   */
  kDifRstmgrResetInfoNdm = (0x1 << 2),
  /**                                                                                                                                                                                                               
   * Device has reset due to a peripheral request. This can be an alert                                                                                                                                             
   * escalation, watchdog or anything else.                                                                                                                                                                         
   */
  kDifRstmgrResetInfoHwReq = (0x1 << 3),
  /**                                                                                                                                                                                                               
   * \internal used to catch missing `_Static_assert` for public variants.                                                                                                                                          
   */
  kDifRstmgrResetInfoLast = kDifRstmgrResetInfoHwReq,
} dif_rstmgr_reset_info_t;
...
```

`dif_rstmgr.c`:
```
// This macro simplifies the `_Static_assert` check to make sure that the                                                                                                                                           
// public reset info register bitfield matches register bits.                                                                                                                                                       
#define RSTMGR_RESET_INFO_CHECK(pub_name, priv_name)                      \
  _Static_assert(                                                         \
      kDifRstmgrResetInfo##pub_name == (0x1 << RSTMGR_RESET_##priv_name), \
      "kDifRstmgrResetInfo" #pub_name " must match the register definition!")

RSTMGR_RESET_INFO_CHECK(Por, INFO_POR);
RSTMGR_RESET_INFO_CHECK(LowPowerExit, INFO_LOW_POWER_EXIT);
RSTMGR_RESET_INFO_CHECK(Ndm, INFO_NDM_RESET);
RSTMGR_RESET_INFO_CHECK(HwReq, INFO_HW_REQ);
...
_Static_assert(
    DIF_RSTMGR_ALERT_INFO_MAX_SIZE == RSTMGR_ALERT_INFO_CTRL_INDEX_MASK,
    "Alert info dump max size has grown, please update the public define!");
...
```

I was on board with this @lenary suggestion, but now I have some doubts. These seem to me like something that can be parameterized differently between the chips (at the moment only EarlGrey), in which case the assertions will fail for on of the instantiations.

What do you think?

silvestrst commented:
Added unittests - `dif_rstmgr_alert_info_dump_read` for now.

silvestrst commented:
I think I might actually prefer `result_t dif_rstmgr_alert_info_dump_read(... *handle, ... *buffer, buffer_len, size_t *words_written)`, that means that I would need to resurrect `...get_size`. Please see the  [comment](https://github.com/lowRISC/opentitan/pull/3867#issuecomment-716438469) above. Yes, this interface is not foolproof, however the caller is in full control of what to pass to the function. The DIF can then return an error if the buffer is not big enough for the entire dump.

I am just double weary now on tying private (definitions that can potentially be parameterized) with the public side of the API (@gkelly @lenary @alphan @mcy @jon-flatley ). What are your thoughts?

alphan commented:
> These seem to me like something that can be parameterized differently between the chips (at the moment only EarlGrey), in which case the assertions will fail for on of the instantiations.
> 
> What do you think?

IMHO it all depends on the actual differences between the chips. For `_dump_read` I don't see a big difference between the two approaches since both of them expect the clients to allocate a fixed size buffer, right? Size of that buffer is currently hard-coded in the DIF header but if we want to generalize it to multiple chips that information must come from somewhere else. Will it be passed to `_init` as a literal or will it come from another auto-generated header? We can also completely decouple the destination buffer size from the hardware but it would require tracking the state of the read operation, which means additional complexity. All of this also depends on how this library will be built for different chips. I think, personally, I would lean towards simplicity unless we have a clear idea about which parts of the hardware can be different and how we want to deal with such changes in all DIFs.

lenary commented:
> I think I might actually prefer `result_t dif_rstmgr_alert_info_dump_read(... *handle, ... *buffer, buffer_len, size_t *words_written)`, that means that I would need to resurrect `...get_size`. Please see the [comment](https://github.com/lowRISC/opentitan/pull/3867#issuecomment-716438469) above. Yes, this interface is not foolproof, however the caller is in full control of what to pass to the function. The DIF can then return an error if the buffer is not big enough for the entire dump.
> 
> I am just double weary now on tying private (definitions that can potentially be parameterized) with the public side of the API (@gkelly @lenary @alphan @mcy @jon-flatley ). What are your thoughts?

It seems like the core of the problem is that we've used a `#define` because we want to use the reset manager buffer size in type definitions, so we don't dynamically allocate. We use a `#define` because we can't use a const global variable, which would allow us to hide the value away from the header. This issue is hit in USBDev as well.

I think my approach would be this:
- We should move to alphan's API as proposed above, for this DIF. This is safer today, and looks like the API we want when we come to parameterize this DIF's implementation against the hardware instance in a more complete manner than we currently do.
- We should avoid, as much as possible, `#defines` in DIF headers, and almost certainly global variables are also a problem when we come to parameterize the hardware - in the long term, we should be doing these queries with functions which take in a DIF instance handle, so the function can query `handle.params`.
- We need to think about the dynamic buffer allocation issue, which we have so-far been punting on by using `#defines` (which can be used in types, e.g. for array sizes, unlike global variables).

I think this comes under the "parameterize the difs" issue we've thought about before. I'll take a look through the current DIFs to highlight which have problematic defines or global variables, and collect those into an issue so we don't lose track of them.

Parameterising only becomes an issue when we have multiple chips - but we want the APIs reasonably stable for shipping the Mask ROM, so we should ensure we're not painting ourselves into a corner with the first API. 

Does this sound reasonable?

silvestrst commented:
^ sounds reasonable to me. I think the current consensus from the offline chat was that we will try to avoid implementing an allocator, but will try to generate relevant defines in the `top_earlgrey.h` so that the DIF consumers could allocate the buffers statically. **This are the thought from a quick offline chat, and has not been explored further in detail!**

silvestrst commented:
In the end to simplify the implementation and avoid implementing additional functionality that we might not need, I have decided to disallow partial dump read, and return an error if the full dump cannot be read.

I have added unittests.

gkelly reviewed (approved):
This looks good to me. I think there's a question here about using general errors for such a specific error result when there's not enough space but I think this is a reasonable way to handle that for now.

silvestrst commented:
> I think there's a question here about using general errors for such a specific error result when there's not enough space

I have added this to my TODOs issue.

Merging now, I think I have addressed all the comments, and this change had enough exposure.
